Abs-Val-Jacobi-type of preconditions/smoothers for PA operators

homeomorfismo opened the pull request:
Cleaner version of #4498.

> The absolute-value L(1)-Jacobi preconditioner is implemented for partially assembled systems. We extend the MFEM interface to support AbsMult for different types of operators. Modifications were required at the integrator level, and we extended some kernels. In general, the strategy was to avoid modifying the existing kernel functions and to ensure that the basis functions (and their derivatives) are positive at the quadrature points.

TODOs: (cf. https://github.com/mfem/mfem/issues/4498#issuecomment-2763126967)
- I need to fix the makefile: https://github.com/mfem/mfem/blob/2e2b8faba5910fa2f11179c8a88d5004bc5f32c4/miniapps/diag-smoothers/makefile#L43-L45
- The makefile should link the meshes to a more convenient directory (like the meshing miniapp)
- I ignored `Array<T>::Apply(std::function<T(T)>)` and just implemented `Array<T>::Abs`
- At some point I thought of using something like `auto fun = useAbs ? [](TA a) { return std::abs(a); } : [](TA a) { return a; }` but I don't know if it would be convenient, efficient, and readable.
- Is there a way to "hide" the `switch` case in the elasticity integrators? I know that the template requires `constexpr` to compile. I haven't tried any ideas yet.
<!--GHEX{"author":"homeomorfismo","assignment":"2025-03-29T19:33:22","editor":"tzanio","id":4769,"reviewers":["tzanio","v-dobrev","vladotomov","pazner"],"merge":"2025-06-26T16:34:45","approval":"2025-06-23T19:33:30"}XEHG-->
<!--GHEXTABLE-->
 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#4769](https://github.com/mfem/mfem/pull/4769) | @homeomorfismo | @tzanio | @tzanio + @v-dobrev + @vladotomov + @pazner | 3/29/25 | 6/23/25 | 6/26/25 | |
<!--ELBATXEHG-->
<details>
<summary>PR Checklist</summary>
    
- [ ] Code builds.
- [ ] Code passes `make style`.
- [ ] Update `CHANGELOG`:
    - [ ] Is this a new feature users need to be aware of? New or updated example or miniapp?
    - [ ] Does it make sense to create a new section in the `CHANGELOG` to group with other related features?
- [ ] Update `INSTALL`:
    - [ ] Had a new optional library been added? If so, what range of versions of this library are required? (*Make sure the external library is compatible with our BSD license, e.g. it is not licensed under GPL!*)
    - [ ] Have the version ranges for any required or optional libraries changed?
    - [ ] Does `make` or `cmake` have a new target?
    - [ ] Did the requirements or the installation process change? *(rare)*
- [ ] Update continuous integration server configurations if necessary (e.g. with new version requirements for each of MFEM's dependencies)
    - [ ] `.github`
    - [ ] `.appveyor.yml`
- [ ] Update `.gitignore`:
    - [ ] Check if `make distclean; git status` shows any files that were generated from the source by the project (not an IDE) but we don't want to track in the repository.
    - [ ] Add new patterns (just for the new files above) and re-run the above test.
- [ ] New examples:
    - [ ] All sample runs at the top of the example source file work.
    - [ ] Update `examples/makefile`:
      - [ ] Add the example code to the appropriate `SEQ_EXAMPLES` and `PAR_EXAMPLES` variables.
      - [ ] Add any files generated by it to the `clean` target.
      - [ ] Add the example binary and any files generated by it to the top-level `.gitignore` file.
    - [ ] Update `examples/CMakeLists.txt`:
      - [ ] Add the example code to the `ALL_EXE_SRCS` variable.
      - [ ] Make sure `THIS_TEST_OPTIONS` is set correctly for the new example.
   - [ ] List the new example in `doc/CodeDocumentation.dox`.
   - [ ] If new examples directory (e.g.`examples/pumi`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
      - [ ] Update or add example-specific documentation, see e.g. the `src/examples.md`.
      - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
      - [ ] In `examples.md`, list the example under the appropriate categories, add new categories if necessary.
      - [ ] Add a short description of the example in the "Extensive Examples" section of `features.md`.
- [ ] New miniapps:
   - [ ] All sample runs at the top of the miniapp source file work.
   - [ ] Update top-level `makefile` and `makefile` in corresponding miniapp directory.
   - [ ] Add the miniapp binary and any files generated by it to the top-level `.gitignore` file.
   - [ ] Update CMake build system:
      - [ ] Update the `CMakeLists.txt` file in the `miniapps` directory, if the new miniapp is in a new directory.
      - [ ] Add/update the `CMakeLists.txt` file in the new miniapp directory.
      - [ ] Consider adding a new test for the new miniapp.
   - [ ] List the new miniapp in `doc/CodeDocumentation.dox`
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), add it to `MINIAPP_SUBDIRS` in the `makefile`.
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
     - [ ] Update or add miniapp-specific documentation, see e.g. the `src/meshing.md` and `src/electromagnetics.md` files.
     - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
     - [ ] The miniapps go at the end of the page, and are usually listed only under a specific "Application (PDE)" category.
     - [ ] Add a short description of the miniapp in the "Extensive Examples" section of `features.md`.
- [ ] New capability:
   - [ ] All new public, protected, and private classes, methods, data members, and functions have full Doxygen-style documentation in source comments. Documentation should include descriptions of member data, function arguments and return values, template parameters, and prerequisites for calling new functions.
   - [ ] Pointer arguments and return values must specify whether ownership is being transferred or lent with the call.
   - [ ] Any new functions should include descriptions of their intended use e.g. for internal use only, user-facing, etc., along with references to example code whenever possible/appropriate.
   - [ ] Consider adding new sample runs in existing examples to highlight the new capability.
   - [ ] Consider saving cool simulation pictures with the new capability in the Confluence gallery (LLNL only) or submitting them, via pull request, to the gallery section of the `mfem/web` repo.
   - [ ] If this is a major new feature, consider mentioning it in the short summary inside `README` *(rare)*.
   - [ ] List major new classes in `doc/CodeDocumentation.dox` *(rare)*.
- [ ] Update this checklist, if the new pull request affects it.
- [ ] Run `make unittest` to make sure all unit tests pass.
- [ ] Run the tests in `tests/scripts`.
- [ ] (LLNL only) After merging:
   - [ ] Update internal tests to include the new features.
</details>

tzanio commented:
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

v-dobrev commented on fem/bilininteg.hpp:
```diff
@@ -113,6 +119,8 @@ class BilinearFormIntegrator : public NonlinearFormIntegrator
        called. */
    void AddMultMF(const Vector &x, Vector &y) const override;
 
+   virtual void AddAbsMultMF(const Vector &x, Vector &y) const;
```
This is not used and can be removed.

v-dobrev commented on fem/bilininteg.hpp:
```diff
@@ -78,9 +78,13 @@ class BilinearFormIntegrator : public NonlinearFormIntegrator
    /// Method for partially assembled action on NURBS patches.
    virtual void AddMultNURBSPA(const Vector&x, Vector&y) const;
 
+   virtual void AddAbsMultNURBSPA(const Vector&x, Vector&y) const;
```
This is not used and can be removed.

v-dobrev commented on fem/bilininteg.hpp:
```diff
@@ -121,6 +129,8 @@ class BilinearFormIntegrator : public NonlinearFormIntegrator
        called. */
    virtual void AddMultTransposeMF(const Vector &x, Vector &y) const;
 
+   virtual void AddAbsMultTransposeMF(const Vector &x, Vector &y) const;
```
This is not used and can be removed.

v-dobrev commented on fem/integ/bilininteg_elasticity_kernels.hpp:
```diff
@@ -164,8 +180,8 @@ void ElasticityAddMultPA_(const int nDofs, const FiniteElementSpace &fespace,
-   E_To_Q_Map->PhysDerivatives(x, QVec);
-
+   if (!useAbs) { E_To_Q_Map->PhysDerivatives(x, QVec); }
+   else { E_To_Q_Map->AbsPhysDerivatives(x, QVec); }
```
To avoid the proliferation of `*Abs*` methods to `QuadratureInterpolator`, instead of using `AbsPhysDerivatives` here, this absolute value should be done inside the kernel below. This way, no changes are needed in `QuadratureInterpolator`.

v-dobrev commented on linalg/densemat.hpp:
```diff
@@ -142,29 +142,35 @@ class DenseMatrix : public Matrix
+   void Mult(const Vector &x, real_t *y, bool useAbs = false) const;
 
-   /// Matrix vector multiplication.
+   /// Matrix vector multiplication
```
Unnecessary change

homeomorfismo commented:
Thanks @v-dobrev! I will look into this tomorrow :) Sorry for being AWOL

homeomorfismo commented:
I will take a look at the drivers: there are some redundant things in there that I forgot to remove (related to the L(p,q) smoothers).

I need to take a look more carefully at the current implementation of the derivatives in the Elasticity case (move that to the kernel level) and the proposed tests (I recall that the Kershaw transform destroys the mesh quality, but I am not sure *how* bad it should impact the solvers).

Thanks for the feedback again, I'll be looking into this :)

pazner commented on CHANGELOG:
```diff
@@ -11,6 +11,15 @@
 ===========================
 
+- API addition: in class `Operator`, added virtual functions: `AbsMult`,
+  `AbsMultTranspose`, `AddAbsMult`, and `AddAbsMultTranspose`.
```
Were the `AddAbsMult*` versions removed since this was written?

pazner commented on fem/bilinearform_ext.hpp:
```diff
@@ -54,6 +54,11 @@ class BilinearFormExtension : public Operator
 
+   void AbsMult(const Vector &x, Vector &y) const override
+   {
+      MFEM_ABORT("AbsMult not implemented for this assembly level!")
```
Do we need this if `Operator::AbsMult` already errors out?

pazner commented on fem/restriction.hpp:
```diff
@@ -171,6 +177,14 @@ class FaceRestriction : public Operator
 
+   virtual void MultUnsigned(const Vector &x, Vector &y) const
+   {
+      MFEM_ABORT("MultUnsigned not implemented yet!");
```
Not quite accurate since derived class `ConformingFaceRestriction` overrides this.

If `AbsMult` is going to be the standard terminology, why not define `MultUnsigned` to call `AbsMult` (for which the default base class implementation will simply error), and then you can override `ConformingFaceRestriction::AbsMult`?

pazner commented on fem/restriction.cpp:
```diff
@@ -653,7 +653,8 @@ ConformingFaceRestriction::ConformingFaceRestriction(
 { }
 
-void ConformingFaceRestriction::Mult(const Vector& x, Vector& y) const
+void ConformingFaceRestriction::MultInternal(const Vector& x, Vector& y,
+                                             const bool useAbs) const
```
One option here (probably not too important) is to make `useAbs` a template parameter, than will avoid the runtime check at every iteration. Likely makes little to no difference though…

pazner commented on fem/integ/bilininteg_diffusion_pa.cpp:
```diff
@@ -164,6 +164,45 @@ void DiffusionIntegrator::AssemblePatchPA(const int patch,
+   {
+      MFEM_ABORT("Ceed AbsMult not implemented yet");
+   }
+   Vector abs_pa_data(pa_data);
+   abs_pa_data.Abs();
+
+   Array<real_t> absB(maps->B);
+   Array<real_t> absG(maps->G);
+   Array<real_t> absBt(maps->Bt);
+   Array<real_t> absGt(maps->Gt);
+
+   absB.Abs();
+   absG.Abs();
+   absBt.Abs();
+   absGt.Abs();
```
This logic is basically going to be the same for most PA integrators — taking the absolute value of B, G, Bt, Gt, and the "PA data" (i.e. the `D` matrix). Can we come up with an abstraction that will do this as a fallback?

pazner commented on fem/quadinterpolator.cpp:
```diff
@@ -540,34 +541,58 @@ void QuadratureInterpolator::Mult(const Vector &e_vec,
+   DofToQuad abs_maps;
+   if (useAbs)
+   {
+      abs_maps.FE = maps.FE;
+      abs_maps.IntRule = maps.IntRule;
+      abs_maps.mode = maps.mode;
+      abs_maps.ndof = maps.ndof;
+      abs_maps.nqpt = maps.nqpt;
+      abs_maps.B = maps.B;
+      abs_maps.Bt = maps.Bt;
+      abs_maps.G = maps.G;
+      abs_maps.Gt = maps.Gt;
```
I think `abs_maps = maps;` will work here instead

pazner replied:
Since similar logic is performed in the integrators, we might want
`DofToQuad DofToQuad::Abs() const` that does this.

pazner commented on fem/quadinterpolator.hpp:
```diff
@@ -143,6 +148,8 @@ class QuadratureInterpolator
        @a e_vec at quadrature points. */
    void PhysDerivatives(const Vector &e_vec, Vector &q_der) const;
 
+   void AbsPhysDerivatives(const Vector &e_vec, Vector &q_der) const;
```
Doxygen comment needed here

v-dobrev replied:
Note: this is one of the things I recommended to be removed.

pazner commented on linalg/operator.cpp:
```diff
@@ -645,25 +645,106 @@ void ConstrainedOperator::ConstrainedMult(const Vector &x, Vector &y,
    }
 }
 
+void ConstrainedOperator::ConstrainedAbsMult(const Vector &x, Vector &y,
```
The implementation should be merged with `ConstrainedOperator::ConstrainedMult` since it basically follows the same pattern

pazner commented on fem/bilinearform_ext.cpp:
```diff
@@ -558,11 +559,13 @@ void PABilinearFormExtension::Mult(const Vector &x, Vector &y) const
          else
          {
-            integrators[i]->AddMultPA(x, y);
+            if (!useAbs) { integrators[i]->AddMultPA(x, y); }
+            else { integrators[i]->AddAbsMultPA(x, y); }
```
Nitpick: better to use `if (useAbs)` rather than the negation

pazner commented on fem/bilinearform_ext.cpp:
```diff
@@ -571,14 +574,30 @@ void PABilinearFormExtension::Mult(const Vector &x, Vector &y) const
-         elem_restrict->Mult(x, localX);
+         auto H1elem_restrict =
+            dynamic_cast<const ElementRestriction*>(elem_restrict);
+         if (H1elem_restrict && useAbs)
+         {
+            H1elem_restrict->AbsMult(x, localX);
+         }
```
Why only in the `H1elem_restrict` case?

Also, I think the name `H1elem_restrict` isn't quite accurate — this will be non-null for anything except L2? But also, for L2, the `AbsMult` should be the same as `Mult` since there are no signs? So maybe `AbsMult` should default to `Mult` in the base class `ElementRestrictionOperator`

pazner commented on fem/bilinearform_ext.cpp:
```diff
@@ -1010,41 +1048,81 @@ void EABilinearFormExtension::Assemble()
    }
    // Apply the Element Matrices
    {
+      Vector abs_ea_data(ea_data.Size());
```
Don't need to allocate if `useAbs` is false.
```suggestion
      Vector abs_ea_data;
```

pazner commented on fem/bilinearform_ext.cpp:
```diff
@@ -1053,7 +1131,9 @@ void EABilinearFormExtension::Mult(const Vector &x, Vector &y) const
    const int iFISz = intFaceIntegrators.Size();
    if (int_face_restrict_lex && iFISz>0)
    {
+      MFEM_ASSERT(!useAbs, "AbsMult not implemented with Face integrators!")
```
```suggestion
      MFEM_VERIFY(!useAbs, "AbsMult not implemented with Face integrators!")
```

pazner reviewed (commented):
Nice work on this PR @homeomorfismo!

Haven't finished a complete review, but I've left some relatively minor feedback for now.

homeomorfismo commented:
> * `abs-l1-jacobi`:
>   
>   * the last sample run crashes:
>     ```
>     mpirun -np 4 ./abs-l1-jacobi -m ../../data/beam-tet.mesh -Ky 0.5 -Kz 0.5
>     ```
I think Kershaw transformation requires full assembly, I haven't fully checked. (I should have been more careful writing the examples, my bad)

homeomorfismo commented:
From https://github.com/mfem/mfem/pull/4769#issuecomment-2822389336 @v-dobrev's comments 

> Here are some issues and requests for changes in the new miniapps:
> 
> * `abs-l1-jacobi`:
>   
>   * the last sample run crashes:
>     ```
>     mpirun -np 4 ./abs-l1-jacobi -m ../../data/beam-tet.mesh -Ky 0.5 -Kz 0.5
>     ```

Mesh `beam-tet.mesh` does not have quad elements. I changed the example.

>   * add device sample runs

I tested this on [ORCA](https://orca.pdx.edu/) with CUDA. This was added to the docstrings.

>   * address todo or remove it:
>     ```
>      TODO(Gabriel): Elasticity not implemented yet for partial assembly
>     ```

Elasticity implementation was removed.

>   * no need to have separate options for `LEGACY` and `LEGACYFULL` -- they are the same

The interface was reduced.

>   * fix the implementation of `AssembleElementLpqJacobiDiag` or remove it.

We kept this, we copy the element matrix, compute its `p` power, and perform classical
multiplication.

> * `mg-abs-l1-jacobi`:
>   
>   * the second sample run is very slow:
>     ```
>      mpirun -np 4 ./mg-abs-l1-jacobi -s 1 -i 3
>     ```

I think this corresponded to Maxwell, which was removed for MG as it requires a specific
preconditioner.

>   * the third sample run is relatively slow for multigrid:
>     ```
>      mpirun -np 4 ./mg-abs-l1-jacobi -m ../meshing/icf.mesh -f 0.5
>     ```

No comments. I recall scale of the mesh itself was different, thus I decided to use a
different frequency for the manufactured solution.

>   * the last sample run crashes:
>     ```
>      mpirun -np 4 ./mg-abs-l1-jacobi -m ../../data/beam-tet.mesh -Ky 0.5 -Kz 0.5
>     ```

Id. See above.

>   * add device sample runs

Id. See above.

>   * no need to have separate options for `LEGACY` and `LEGACYFULL` -- they are the same.

Id. See above.

---

Could you give it a second review 🥺  @v-dobrev @tzanio @pazner

homeomorfismo commented on fem/integ/bilininteg_mass_pa.cpp:
```diff
@@ -199,10 +199,37 @@ void MassIntegrator::AddMultPA(const Vector &x, Vector &y) const
+   {
+      Vector abs_pa_data(pa_data);
+      abs_pa_data.Abs();
+      Array<real_t> absB(maps->B);
+      Array<real_t> absBt(maps->Bt);
+      absB.Abs();
+      absBt.Abs();
```
I will update this

homeomorfismo commented on fem/integ/bilininteg_vectorfemass_pa.cpp:
```diff
@@ -313,6 +313,129 @@ void VectorFEMassIntegrator::AddMultPA(const Vector &x, Vector &y) const
+   Vector abs_pa_data(pa_data);
+   abs_pa_data.Abs();
+
+   Array<real_t> absBo(mapsO->B);
+   Array<real_t> absBc(mapsC->B);
+   Array<real_t> absBto(mapsO->Bt);
+   Array<real_t> absBtc(mapsC->Bt);
+   Array<real_t> absBto_t(mapsOtest->Bt);
+   Array<real_t> absBtc_t(mapsCtest->Bt);
+
+   absBo.Abs();
+   absBc.Abs();
+   absBto.Abs();
+   absBtc.Abs();
+   absBto_t.Abs();
+   absBtc_t.Abs();
```
I will update this too -- using `maps->Abs()` instead of manual copies to each array.

v-dobrev replied:
I didn't update these to save some redundant `Array` allocations and `Array::Abs` calls -- there are a total of 4 `Array`s copied (and `Abs` applied to them) for each `DofToQuad::Abs` call. If we use 4 such calls (for `mapsO`, `mapsC`, `mapsOtest`, and `mapsCtest`) then we end up copying 16 arrays while the current code uses just 6. On the other hand these `Array`s are relatively small, so the overhead of using `DofToQuad::Abs` is not that big.

homeomorfismo replied:
I got carried away. I think the last commit can be dropped directly (sorry about that!). I was thinking about that reason but I assumed aiming for similar structure. I don't know what is preferred (yet it would be easier to just drop the last commit). (Next time I will ask and wait)

v-dobrev commented on fem/integ/bilininteg_vectorfemass_pa.cpp:
```diff
@@ -323,19 +323,8 @@ void VectorFEMassIntegrator::AddAbsMultPA(const Vector &x, Vector &y) const
-   absBto_t.Abs();
-   absBtc_t.Abs();
+   auto abs_mapsO = mapsO->Abs();
+   auto abs_mapsC = mapsC->Abs();
```
Also, just to be clear: in the changes below, this commit replaces `absBto_t` and `absBtc_t` with `abs_mapsO.Bt` and `abs_mapsC.Bt`, respectively, which is generally incorrect.

homeomorfismo replied:
Oh gosh, I misread that. Thanks @v-dobrev!

homeomorfismo commented:
> * `mpirun -np 4 abs-l1-jacobi -t 1e5 -ni 100 -vis`
> * `mpirun -np 4 mg-abs-l1-jacobi -t 1e5 -ni 100 -vis`
These were a typo. I replaced them with `1e-5`.

I addressed the removed integrators, and changed the `visualization` flag to `true`. Thanks @v-dobrev!

v-dobrev commented:
Thanks for the updates @homeomorfismo!

I noticed that now the following two sample runs are actually identical:
* `mpirun -np 4 mg-abs-l1-jacobi`
* `mpirun -np 4 mg-abs-l1-jacobi -s 1 -i 1`

because `-s 1 -i 1` are also the default values.

Should we change the second one in some way, or just remove it?

homeomorfismo commented:
> I noticed that now the following two sample runs are actually identical:
> 
> * `mpirun -np 4 mg-abs-l1-jacobi`
> * `mpirun -np 4 mg-abs-l1-jacobi -s 1 -i 1`
> 
> because `-s 1 -i 1` are also the default values.
> 
> Should we change the second one in some way, or just remove it?

I changed the example to use SLI and the mass system (L2-projection), because otherwise it will be quite slow with the current tolerance.

v-dobrev commented:
Merged in `next` for testing...
